shared/js/smart-screen/cards/folder.js

```diff
@@ -44,51 +44,102 @@
+    return this._cardsInFolder;
+  };
+
+  // get index of card in folder
+  Folder.prototype._indexOfCard = function folder_indexOfCard(query) {
+    return this._cardsInFolder.indexOf(this.findCard(query));
+  };
+
+  // TODO: This is almost the same as CardManager.findCardFromCardList()
+  // We should merge and generalize them into one
+  // There are three types of query:
+  // 1. query by cardId
+  // 2. query by manifestURL and optionally launchURL
+  // 3. query by cardEntry (i.e. serialized card)
+  Folder.prototype.findCard = function folder_findCard(query) {
+    var found;
+    this._cardsInFolder.some(function(card, index) {
+      if (card.cardId === query.cardId) {
+        found = card;
+        return true;
+      } else if (query.manifestURL && card.nativeApp &&
+          card.nativeApp.manifestURL === query.manifestURL) {
+        // if we specify launchURL in query, then we must compare
+        // launchURL first
+        if (query.launchURL) {
+          if (card.launchURL === query.launchURL) {
+            found = card;
+            return true;
+          }
         } else {
-          return false;
+          found = card;
+          return true;
         }
-      });
-    }
-    return indexOfFound;
+      } else if (query.cardEntry) {
+        // XXX: this could be bad at performance because we serialize card
+        // in every loop. We might need improvement on this query.
+        if (JSON.stringify(card.serialize()) ===
+            JSON.stringify(query.cardEntry)) {
+          found = card;
+          return true;
+        }
+      }
+    });
+    return found;
   };
 
-  Folder.prototype._isInFolder = function folder_isInFolder(cardInSearch) {
-    if (cardInSearch) {
-      return this._findInFolder(cardInSearch) > -1;
+  Folder.prototype._isInFolder = function folder_isInFolder(card) {
+    if (card) {
+      return this._indexOfCard(card) > -1;
     } else {
       return false;
     }
   };
 
+  Folder.prototype._dirty = function folder_dirty() {
+    if (this.state !== Folder.STATES.DETACHED) {
+      this.state = Folder.STATES.DIRTY;
+    }
+    this.fire('folder-changed', this);
+  };
+
   Folder.prototype.addCard = function folder_addCard(card) {
     // We don't support folder in folder
     if (!this._isInFolder(card) && !(card instanceof Folder)) {
-      this.cardsInFolder.push(card);
-      if (this.state !== Folder.STATES.DETACHED) {
-        this.state = Folder.STATES.DIRTY;
-      }
-      this.fire('folder-changed', this);
+      this._cardsInFolder.push(card);
+      this._dirty();
     }
   };
 
   Folder.prototype.removeCard = function folder_removeCard(card) {
-    var index = this._findInFolder(card);
+    var index = this._indexOfCard(card);
     if (index > -1) {
-      this.cardsInFolder.splice(index, 1);
-      if (this.state !== Folder.STATES.DETACHED) {
-        this.state = Folder.STATES.DIRTY;
-      }
-      this.fire('folder-changed', this);
```

Comment:
Sure.  But I'll add `card-removed` event instead of replace `folder-changed` event. Because we need `folder-changed` event to let card manager to write folder back to data store.